File: exercises/01-cfd/code/c-ctypes/jacobi.c
/* 
 * Jacobi iteration
 *
 * m+2     m is physical number of points x-direction
 * n+2     n is physical number of points y-direction
 * niter   number of jacobi iterations
 * psi     stream function expected to be of extent (m+2)*(n+2)
 *
 * Returns 0 on success.
 */

#include <stdlib.h>

int jacobi(int m, int n, int niter, double * psi) {

  int iter;
  int i, j;
  double * tmp = NULL;

  /* Temporary space of extent m*n for the update */

  tmp = (double *) calloc(m*n, sizeof(double));
  if (tmp == NULL) return -1;

  for (iter = 0; iter < niter; iter++) {
    /* Compute differences in true domain 1..m 1..n */
    for (i = 1; i <= (m-2); i++) {
      for (j = 1; j <= (n-2); j++)  {
	int im1 = (i-1)*n + j;
	int ip1 = (i+1)*n + j;
	int jm1 = i*n + j - 1;
	int jp1 = i*n + j + 1;
	tmp[i*n + j] = 0.25*(psi[im1] + psi[ip1] + psi[jm1] + psi[jp1]);
      }
    }

    /* Update */
    for (i = 1; i <= (m-2); i++)  {
      for (j = 1; j <= (n-2); j++)  {
	psi[i*n + j] = tmp[i*n + j];
      }
    }
  }

  free(tmp);

  return 0;
} 

File: exercises/01-cfd/solutions/numpy_index/jacobi.py
# 
# Jacobi routine for CFD calculation
#
import numpy as np

def jacobi(niter, psi):

    (m, n) = psi.shape
    m = m - 2
    n = n - 2

    tmp = np.zeros((m+2, n+2))
    for iter in range(niter):
        # Use index notation and offsets to compute the stream function
        tmp[1:m+1,1:n+1] = 0.25 * (psi[2:m+2,1:n+1]+psi[0:m,1:n+1]+psi[1:m+1,2:n+2]+psi[1:m+1,0:n])

        # Update psi
        np.copyto(psi[1:m+1,1:n+1], tmp[1:m+1,1:n+1])

